### PR TITLE
[7/X][Pipeline] pipeline user facing build function

### PR DESCRIPTION
**Pull request by jiaodong**
## Diff Summary

- Renamed `_bind()` and `_apply_functional()` as public API.
- Added a public facing `build()` function in `pipeline/api.py`
  - With integration test using same complex patterns of existing deployment test and ensure output is the same.
- Added JSON serialization for InputAttributeNode  
- Added utils function to find and assert existence of `PipelineInputNode` upon calling `build()`
  - With unit test
- Added assertions on FunctionNode, ClassNode, DeploymentNode and PipelineInputNode to always ensure fully qualified name upon generating import_path
  - With unit tests
- Moved common DAG pattern and test modules to `resources` subfolder for code reuse
- Added `test_build.py` to serve CI. 

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by edoakes — COMMENTED** [review]
@jiaodong I think we should enforce the importability in the `to_yaml` / `to_json` step (in serve app) instead of here. Otherwise, people can't fully do local testing without making things importable?

**Comment by jiaodong** [comment]
To be clear we want 

```
app = serve.pipeline.build(ray_dag) # This should work

handle = app.deploy() 
ray.get(handle.remote(data)) # This should also work

deployment_yaml = app.to_yaml() # This should fail if imports are inline 
```

however this PR is written without the `app` part landed yet and just returned a list of deployments, so we need to move some pieces to app level once it landed. Unit tests in this PR are to ensure our assertion works as expected at JSON serialization step.

Once app PR is landed I think the easiest thing is to pass an enforcement flag to our JSON encoder / object hook to enforce it on each DAGNode's serde call.

File: python/ray/serve/pipeline/deployment_node.py [hunk 1]
```diff
@@ -169,7 +169,19 @@ def get_import_path(self):
...
+            "Class used in DAG should not be in-line defined when exporting"
+            "import path for deployment. Please ensure it has fully "
+            "qualified name with valid __module__ and __qualname__ for "
+            "import path, with no __main__ or <locals>. \n"
+            f"Current import path: {import_path}"
+        )
+        assert "__main__" not in import_path, error_message
+        assert "<locals>" not in import_path, error_message
+
+        json_dict["import_path"] = import_path
```
Comment (edoakes) [inline python/ray/serve/pipeline/deployment_node.py:184]: are you removing this given the decision to do it at app level?

also, does this detection cover the case where a function returns a class?

**Review by edoakes — APPROVED** [review]

File: python/ray/serve/pipeline/deployment_node.py (hunk 1 above)
Comment (jiaodong) [reply python/ray/serve/pipeline/deployment_node.py:184]: I think the easiest thing is to pass `enforce` flag to each relevant class' to_json implementation, i think leaving it per class works better for localizing implementation, ex: ClassNode and FunctionNode uses `import_path` str while `PipelineInputNode` enforces `preprocessor_import_path`.

Then we only need JSON serialization enforcement on the `get_ingress_deployment` call which is not needed for local testing; And only enforce import path on `to_yaml()` so JSON serializable serve dag with inline classes can still be locally tested with HTTP.

 > also, does this detection cover the case where a function returns a class?

you mean a top-level function that returns a class as value ? Let me add a test for this.

Comment (jiaodong) [reply python/ray/serve/pipeline/deployment_node.py:184]: added a test with class factory and verified it fails json serde.

**Comment by jiaodong** [comment]
Locally verified all ray dag tests and serve pipeline tests passed, which are the only parts this PR touched.

Only exception being renaming _bind to bind that effected Actor / Function implementation, with very limited # of lines changed.

**Pull request merged**